core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java

```diff
@@ -158,25 +120,26 @@ public void onFailure(Throwable t) {
             } else {
                 logger.trace("no need to schedule reroute due to delayed unassigned, next_delay_setting [{}], registered [{}]", nextDelaySetting, registeredNextDelaySetting);
             }
-        } else {
-            FutureUtils.cancel(scheduledRoutingTableFuture);
-            scheduledRoutingTableFuture = null;
         }
     }
 
-    private void reroute() {
+    // visible for testing
+    long getRegisteredNextDelaySetting() {
+        return this.registeredNextDelaySetting;
+    }
+
+    // visible for testing
+    void performReroute(String reason) {
         try {
-            if (!routingTableDirty) {
-                return;
-            }
             if (lifecycle.stopped()) {
                 return;
             }
             if (rerouting.compareAndSet(false, true) == false) {
-                logger.trace("already has pending reroute, ignoring");
+                logger.trace("already has pending reroute, ignoring {}", reason);
```

Comment:
Just an idea how about making rerouting an AtomicReference<String> where a non null value signals the currently next reroute task name? this way we can log here logger.trace("already has pending reroute due to [{}]. ignoring [{}]")/